File: packages/cloudflare/src/tools/react.ts
/**
 * React Tool - Add/remove emoji reactions on messages
 */

import { SlackClient, SlackAPIError } from "../slack-client";

interface ReactOptions {
  channel: string;
  messageTs: string;
  emoji: string;
  remove?: boolean;
}

interface ReactResult {
  ok: boolean;
  action?: string;
  emoji?: string;
  channel?: string;
  message_ts?: string;
  channel_name?: string;
  error?: string;
}

/**
 * Add or remove an emoji reaction on a message.
 * This does NOT mark messages as read.
 */
export async function react(
  client: SlackClient,
  options: ReactOptions
): Promise<ReactResult> {
  const { channel, messageTs, emoji, remove = false } = options;

  // Normalize emoji name - strip colons if provided
  const emojiName = emoji.replace(/^:|:$/g, "");

  try {
    if (remove) {
      await client.removeReaction(channel, messageTs, emojiName);
    } else {
      await client.addReaction(channel, messageTs, emojiName);
    }

    const result: ReactResult = {
      ok: true,
      action: remove ? "removed" : "added",
      emoji: emojiName,
      channel,
      message_ts: messageTs,
    };

    // Try to get channel name for context
    try {
      const conv = await client.getConversationInfo(channel);
      if (conv.is_im && conv.user) {
        const user = await client.getUserInfo(conv.user);
        const displayName = user.display_name || user.real_name || user.name;
        result.channel_name = `DM with @${displayName}`;
      } else {
        result.channel_name = conv.name || channel;
      }
    } catch {
      result.channel_name = channel;
    }

    return result;
  } catch (error) {
    const apiError = error as SlackAPIError;
    return {
      ok: false,
      error: `Failed to ${remove ? "remove" : "add"} reaction: ${apiError.error || apiError.message}`,
      emoji: emojiName,
      channel,
      message_ts: messageTs,
    };
  }
}

File: packages/python/src/slack_stealth_mcp/tools/react.py
"""Tool for adding/removing emoji reactions to messages."""

from __future__ import annotations

from typing import Any

from ..slack_client import SlackAPIError
from ..workspace import WorkspaceManager


async def react(
    manager: WorkspaceManager,
    channel: str,
    message_ts: str,
    emoji: str,
    remove: bool = False,
    workspace: str | None = None,
) -> dict[str, Any]:
    """Add or remove an emoji reaction on a message.

    Args:
        manager: Workspace manager
        channel: Channel ID containing the message
        message_ts: Timestamp of the message to react to
        emoji: Emoji name without colons (e.g., "thumbsup", not ":thumbsup:")
        remove: If True, remove the reaction instead of adding
        workspace: Workspace name (uses default if not specified)

    Returns:
        Dictionary with reaction details
    """
    client = manager.get_client(workspace)

    # Normalize emoji name - strip colons if provided
    emoji_name = emoji.strip(":")

    try:
        if remove:
            await client.remove_reaction(channel, message_ts, emoji_name)
            action = "removed"
        else:
            await client.add_reaction(channel, message_ts, emoji_name)
            action = "added"

        result: dict[str, Any] = {
            "ok": True,
            "action": action,
            "emoji": emoji_name,
            "channel": channel,
            "message_ts": message_ts,
        }

        # Try to get channel name for context
        try:
            conv = await client.get_conversation_info(channel)
            if conv.is_im and conv.user:
                user = await client.get_user_info(conv.user)
                result["channel_name"] = f"DM with @{user.display}"
            else:
                result["channel_name"] = conv.display_name
        except Exception:
            result["channel_name"] = channel

        return result

    except SlackAPIError as e:
        return {
            "ok": False,
            "error": f"Failed to {('remove' if remove else 'add')} reaction: {e.error}",
            "emoji": emoji_name,
            "channel": channel,
            "message_ts": message_ts,
        }
